### ISSUE TITLE
Projectiles which lost their origin vehicle generate debug prompts

### ISSUE THREAD

**Issue by emptyVoid** [OpenApoc/OpenApoc#876]
![image](https://user-images.githubusercontent.com/13639229/82362050-67387c80-9a14-11ea-8db4-6e13322cd29f.png)
![image](https://user-images.githubusercontent.com/13639229/82362063-6c95c700-9a14-11ea-88ea-320d47cc7cc7.png)
These prompts get generated every tick till the projectile hits something or expires.

Here are the events that led to the issue:
1. Dimension probe entered the alien dimension.
2. UFO close to the portal started firing at the dimension probe.
3. Dimension probe left the alien dimension.
4. The same UFO invaded the human dimension and got destroyed.
5. A group of XCOM vehicles re-entered the alien dimension.

It looks like time does not flow in the alien dimension without an observer.

The repro save game (just send some vehicles to the portal):
[save_issue-865.zip](https://github.com/OpenApoc/OpenApoc/files/4652320/save_issue-865.zip)

**Comment by JonnyH** [comment]
I think we may need a WeakStateRef (or similar) - that has semantics similar to a weak_ptr<> - IE you 'lock' it that may return a valid StateRef.

This can then be used for links where the target may be destroyed (e.g. 'who fired this' links for experience - if the firer died then that experience can safely be dropped on the floor)

Though we probably still need a "firing organisation" attached to projectiles - so they still blame you for knocking down their buildings even if the firer has disappeared.

**Comment by emptyVoid** [comment]
Both approaches sound good to me.

Though the underlying problem would still be unresolved -- ~~the spice~~ time must flow, even when not observed. Projectiles waiting for player to return to Alien Dimension is kind of weird.

**Comment by FilmBoy84** [comment]
Time in Apoc is certainly an interesting one

The fact that an X-COM craft can enter the alien dimension at any time of the day; but does not allow the player to return to that time of the day until 00:00 to take their actions confuses a lot of people

In the OG, if you enter at 16:00, you will see the alien dimension at 00:00, but the time in the alien dimension will be back to 16:00 (as it should be)

This means that for the cityscape side, there needs to be a log of where projectiles were at 00:00 so when the player is teleported to the Alien dimension at 00:00 to play at 16:00 the previous day once those actions are completed, the cityscape will show the projectiles in the positions that they should be for 00:00

However, it doesn't quite work the same way for projectiles in the alien dimension because whatever action you take there will have a 24-hour gap before you can return (which means those projectiles will have crashed or dissipated)

The OG alien dimension mechanics also do not allow for multiple X-COM incursions at different times of the day - if you send in an X-COM craft at 14:00 and then another at 16:00 both will appear with a claimed 14:00 entry when the user passes 00:00

Depending on how we store projectiles in relation to time we will affect future ability to accurately allow the user to travel between dimensions (especially if mods add multiple ones later)

We should probably consider how we want to do this going forward and where retaining this information is actually useful...

**Comment by JonnyH** [comment]
I don't think running the full simulation for the alien dimension all the time is needed - it's a lot of work for something that isn't observed 99% of the time and doesn't really affect gameplay.

We could tweak it so when the last xcom craft leaves, all projectiles are removed and all damages ufos healed - which is probably what the OG did, and would be equivalent for anything but the weirdest situations.

I don't really like the "reversing time" and "take a whole day" nature of the OG alien dimension travel, but I assume forcing the craft to be away for a full day (and thus not available if theres a UFO incursion while they're away or corporation raid) was intended, even if in most playthroughs by the time you get to that point they aren't much trouble.

**Comment by FilmBoy84** [comment]
Agreed, TBH apart from remembering stuff from the cityscape for the jump between screens at 00:00 which obviously, the player needs to return to when they come back from the alien dimension there is no need to save projectiles within the dimension

The ONLY circumstance in which tracking of both nearly all the time would be needed would be a post 1.0 scenario anyway (if a modder wants to add in the instant-jump tech like the dimension shifter and go between the two when they please as the aliens do)

And that would be also best left for a post 1.0 "fix"

If cutting things down to the bare minimum to simulate OG behaviour can be done now, that's the route we should take and deal with the issues of "but mods" when we have the Original Game simulation sorted in terms of mimicked functionality...
